### PR TITLE
Help room: Create a basic AI to answer questions

### PR DESCRIPTION
**Pull request by panpawn**
WIP

### CODE REVIEW

File: chat-plugins/help-ai.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		} else if ((question.includes('staff') || question.includes('mod') || question.includes('room owner')) && (question.includes('abusing') || question.includes('abuse'))) {
+			return `If you feel as though a <strong>room staff</strong> member has abused their authority, you can PM a Room Owner (#) from that room. Do <strong>/roomauth [roomname]</strong> to see if any are currently online (bolded on the list.) If it was a Room Owner that abused you, you can PM a Moderator (@) in the Help room with your claim and evidence.<br />` +
+				`If you feel as though a <strong>global staff</strong> member (moderator, driver) has abused their authority, you can submit a thread in the ${this.otherAdminRequests('Other Admin Requests Forum.')} Please be thorough in your report, explaining clearly what the problem is, how it happened, who it was with and providing as much proof as you currently have available of any wrongdoings. Be assured that ONLY Pokemon Showdown Senior Staff are able to view the post in that forum, and confidentially will be assured. Complaints are taken very seriously, and any false reports will be dealt with harshly.`;
```
Comment (jumbowhales) [inline chat-plugins/help-ai.js]: I don't think that this, the staff, or the pwreset one should be included here, as they're only really valid on main.

Comment (panpawn) [inline chat-plugins/help-ai.js]: I mean, this entire plugin is kind of specific to main.

Comment (panpawn) [inline chat-plugins/help-ai.js]: PS does not come with a room named "Help" by default, just as the other plugins for other rooms assume, so why would this be any different?

**Comment by bumbadadabum** [comment]
This isn't an AI fwiw.

**Comment by bumbadadabum** [comment]
Also, this really messes with the idea that chat plugins should not interfere with the core functionalities of PS and should be able to be removed. I was planning on making changes to the chat plugin API to better facilitate this, so I guess I'll get working on that sometime soon.

**Comment by panpawn** [comment]
>This isn't an AI fwiw.

What makes this not an AI?

**Comment by jumbowhales** [comment]
Is this really necessary when the roomFAQs exist?

**Comment by panpawn** [comment]
>Is this really necessary when the roomFAQs exist?

The roomfaq system does not stop users from asking questions that are covered in it.

The entire idea behind this is so that we give information that is already readily available and make it become more readily available.

**Comment by bumbadadabum** [comment]
> What makes this not an AI?

I don't see any "intelligent" behavior from this plugin besides .includes(), so I wouldn't call it an AI, rather an answering machine.

> Is this really necessary when the roomFAQs exist?

To be honest, outside of the messages showing up in the room, using roomFAQs has quite a few benefits over this implementation, most notably the fact that things aren't hardcoded into the server, and can be updated by anyone, not just a developer, don't automatically get included on other servers too, and don't require a hotpatch to update. This flexibility is worth quite a lot when it comes to features like this.

**Comment by panpawn** [comment]
>This flexibility is worth quite a lot when it comes to features like this.

"This flexibility" can be apart of this system, potentially

**Comment by panpawn** [comment]
>I don't see any "intelligent" behavior from this plugin besides .includes(), so I wouldn't call it an AI, rather an answering machine.

"intelligent", in the forms of AI, is a pretty broad term - I feel this fits that very well

**Comment by Dragotic-PS** [comment]
`this.includes('battling')`

What if the question is "I'm not really into **battling**, can you recommend something else to do here?"
It can't quite judge the context....

**Comment by AustinXII** [comment]
I agree with bumba, although it is a broad term, detecting words used by people is not artificial intelligence imo

**Comment by panpawn** [comment]
That's kind of exactly how AI works.

When you use a voice assistant on your mobile device, you'll ask it things like "what's the weather?", "who was the 44th president of the United States?", "how far is the earth from the moon?"

For all of these questions, the AI picks up on key words to try and find an answer.

This is the same concept. This is an AI, granted, a much more basic and specific one.

**Comment by panpawn** [comment]
>What if the question is "I'm not really into battling, can you recommend something else to do here?"
It can't quite judge the context....

That if statement has since been corrected

Also, for those worried about this possibly not answering someone's question or picking up on a false-positive, I made it so it has a handy button:

**Comment by AustinXII** [comment]
@panpawn most of the time the AI is learning from other users, besides that it's more of a help faq that can detect words instead of commands

**Comment by bumbadadabum** [comment]
What Austin says. Also, if you  ask the question in the help room, like the button says, wouldn't that bring up the same exact window as it did before?

File: chat-plugins/help-ai.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			let buff = '';
+			buff += Chat.html`<strong>Question:</strong> ${question}<br />`;
+			buff += `<strong>Answer:</strong> ${autoReply}`;
+			if (!command) buff += `<br /><br /><button class="button" name="send" value="HELP: ${question}">This did not answer my question; ask the Help room</button>`;
+			return buff;
+		} else {
+			return false;
+		}
+	}
+	includes(text) {
```
Comment (bumbadadabum) [inline chat-plugins/help-ai.js]: Why?

**Comment by panpawn** [comment]
> Also, if you ask the question in the help room, like the button says, wouldn't that bring up the same exact window as it did before?

No

> most of the time the AI is learning from other users

That's usually not the case, actually

File: chat-plugins/help-ai.js (hunk 2 above)
Comment (panpawn) [inline chat-plugins/help-ai.js]: ``this.includes`` is shorter than ``question.includes``, and I'm thinking of making this function smarter later

**Comment by Zarel** [comment]
Can we PLEASE stop nitpicking about what "AI" means? It's a very broad term and it doesn't actually matter if it's being used correctly, here. Focus on the code and its usefulness.

**Comment by panpawn** [comment]
@Zarel - what are your thoughts on this? Does this sound like a good idea? Is this a good approach?

**Comment by Zarel** [comment]
You should pullreq a separate pullreq for adding chatfilter functionality for plugins.

**Comment by panpawn** [comment]
Okay then, https://github.com/Zarel/Pokemon-Showdown/pull/3868 should probably be taken care of before this

**Comment by panpawn** [comment]
Now that that was taken care of, this can be better implemented.

I just updated this to make use of the new brilliantly redone chatfilter feature in this plugin.

**Comment by panpawn** [comment]
bump?

**Comment by AustinXII** [comment]
This should be shown to help staff. I'm
Personally iffy about this feature, but no one will know better than the active help room staffers.

**Comment by Zarel** [comment]
This is in theory a good idea, but as-is I think is the wrong direction to go.

Better ideas:

- Flesh out the report form to answer common questions.
- Do the chatfilter directly, instead of needing an AI to be explicitly invoked with `HELP:`
- If you want an explicitly-invoked AI, maybe tie it to `/faq` or something; it doesn't need custom syntax.

**Pull request closed without merging**